File: src/java/magma/api/stream/RangeHead.java
package magma.api.stream;

import magma.api.option.None;
import magma.api.option.Option;
import magma.api.option.Some;

public class RangeHead implements Head<Integer> {
    private final int extent;
    private int counter = 0;

    public RangeHead(int extent) {
        this.extent = extent;
    }

    @Override
    public Option<Integer> next() {
        if (counter >= extent) return new None<>();

        var value = counter;
        counter++;
        return new Some<>(value);
    }
}

File: src/c/magma/api/stream/RangeHead.c
struct RangeHead implements Head<Integer>{
	int extent;
	int counter=0;
	public RangeHead(int extent){
		this.extent = extent;
	}
	Option<Integer> next(){
		if(this.counter >= this.extent)return new None<>();
		var value=this.counter;
		this.counter++;
		return new Some<>(value);
	}
}
